parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	/// The module configuration trait.
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The origin to control the collective announcements.
+		type AnnouncementOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin to control the collective charter.
+		type CharterOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The maximum number of announcements.
+		#[pallet::constant]
+		type MaxAnnouncementsCount: Get<u32>;
+
+		/// Weights information needed for the pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Errors encountered by the pallet (not a full list).
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// The announcement is not found.
+		MissingAnnouncement,
+		/// Number of announcements exceeds `MaxAnnouncementsCount`.
+		TooManyAnnouncements,
+	}
+
+	/// Events emitted by the pallet.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		/// A new charter has been set.
+		NewCharterSet { cid: Cid },
+		/// A new announcement has been proposed.
+		AnnouncementAnnounced { cid: Cid },
+		/// An on-chain announcement has been removed.
+		AnnouncementRemoved { cid: Cid },
+	}
+
+	/// The collective charter.
+	#[pallet::storage]
+	#[pallet::getter(fn charter)]
+	pub type Charter<T: Config<I>, I: 'static = ()> = StorageValue<_, Cid, OptionQuery>;
+
+	/// The collective announcements.
+	#[pallet::storage]
+	#[pallet::getter(fn announcements)]
+	pub type Announcements<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, BoundedVec<Cid, T::MaxAnnouncementsCount>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Set the collective  charter.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::Cid) of the IPFS document of the collective charter.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_charter())]
+		pub fn set_charter(origin: OriginFor<T>, cid: Cid) -> DispatchResult {
+			T::CharterOrigin::ensure_origin(origin)?;
+
+			Charter::<T, I>::put(&cid);
+
+			Self::deposit_event(Event::<T, I>::NewCharterSet { cid });
+			Ok(())
+		}
+
+		/// Publish an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::Cid) of the IPFS document to announce.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::announce())]
+		pub fn announce(origin: OriginFor<T>, cid: Cid) -> DispatchResult {
```

Comment:
IMO we should have an expiration (either a block number at which it expires, or a number of blocks for which it is valid, i.e. `AtOrAfter`). Then `on_initialize` should clean up announcements when their expiration block comes. Without this, it doubles the work of members to remove announcements (all for the charitable act of cleaning up some state). I think `Announcements` would become something like a map `BlockNumber => BoundedVec<Cid, ...>`. That should give `on_initialize` an easy job: `if Announcements.contains(now) { take }`.